editor/gui/scene_tree_editor.cpp

```diff
@@ -964,26 +964,55 @@ void SceneTreeEditor::_rename_node(Node *p_node, const String &p_name) {
 	String new_name = p_name.validate_node_name();
 
 	if (new_name != p_name) {
-		error->set_text(TTR("Invalid node name, the following characters are not allowed:") + "\n" + String::get_invalid_node_name_characters());
-		error->popup_centered();
-
-		if (new_name.is_empty()) {
-			item->set_text(0, p_node->get_name());
-			return;
+		String text = TTR("Invalid node name, the following characters are not allowed:") + "\n" + String::get_invalid_node_name_characters();
+		if (error->is_visible()) {
+			if (!error->get_meta("invalid_character", false)) {
+				error->set_text(error->get_text() + "\n\n" + text);
+				error->set_meta("invalid_character", true);
+			}
+		} else {
+			error->set_text(text);
+			error->set_meta("invalid_character", true);
+			error->set_meta("same_unique_name", false);
+			error->popup_centered();
 		}
+	}
 
-		item->set_text(0, new_name);
+	// Trim leading/trailing whitespace to prevent node names from containing accidental whitespace, which would make it more difficult to get the node via `get_node()`.
+	new_name = new_name.strip_edges();
+	if (new_name.is_empty()) {
+		// If name is empty, fallback to class name.
+		if (GLOBAL_GET("editor/naming/node_name_casing").operator int() != NAME_CASING_PASCAL_CASE) {
+			new_name = Node::adjust_name_casing(p_node->get_class());
+		} else {
+			new_name = p_node->get_class();
+		}
 	}
 
 	if (new_name == p_node->get_name()) {
 		if (item->get_text(0).is_empty()) {
 			item->set_text(0, new_name);
 		}
+		return;
+	}
 
+	// we previously made sure name is not the same as current name so that it wonn't complain about already used unique name when not changing name
```

Comment:
```suggestion
	// We previously made sure name is not the same as current name so that it won't complain about already used unique name when not changing name.
```

Reply:
Poke @ajreckof, would be good to fix this before merging.